Feature/1.7/buffer fixes

KristianIvarsson opened the pull request:
Fixes for issue #302 and some other minor return code fixes

KristianIvarsson committed:
buffer: zero sized buffers is not found in lookup table

Make sure to reserve at least one byte in order to get a valid address to the buffer

Resolves #302

KristianIvarsson committed:
buffer: some buffers returned codes indicating an invalid buffer was used

More accurate codes are now returned and some additional logging is done

OlofGullnas commented on middleware/buffer/source/field.cpp:
```diff
@@ -97,7 +97,7 @@ namespace casual
                   if( cursor > end)
-                     common::code::raise::error( common::code::xatmi::argument, "buffer is comprised");
+                     throw std::invalid_argument{ "buffer is comprised"};
                }
```
Is "comprised" the correct word? The word "comprise" with past sense "comprised" has the meaning "consists of" eller "be made up of". Is it the word "compromised" that shold be used?  Or perhaps even "corrupt".

KristianIvarsson replied:
This error message have been there for many years and in practice it isn't even used for logging or anything so it can even be removed entierly

OlofGullnas commented on middleware/buffer/source/field.cpp:
```diff
@@ -170,7 +170,7 @@ namespace casual
                {
-                  Trace trace{ "buffer::field::Allocator::adopt"};
+                  //Trace trace{ "buffer::field::Allocator::adopt"};
                   auto& buffer = allocator_base::emplace_back( std::move( payload), inbound);
```
Should Trace be commented? If this is intentional, should the line be removed instead?

<edit>
I noticed that there are several "//Trace ...." later in the file, so perhaps this is according to "style" in this file. Statements provided in commented form for easy turn on in case needed, but commented because they produces to much "noise" if turned on under normal circumstances.

KristianIvarsson replied:
I guess your analysis is correct

OlofGullnas commented on middleware/buffer/source/field.cpp:
```diff
@@ -247,7 +247,7 @@ namespace casual
                   if( error.code() == common::code::xatmi::argument)
-                     return CASUAL_FIELD_INVALID_HANDLE;
+                     return CASUAL_FIELD_INVALID_HANDLE; 
 
```
Remove space at end of line? reduces noise in diff...

OlofGullnas commented on middleware/buffer/source/octet.cpp:
```diff
@@ -98,8 +100,12 @@ namespace casual
+                     return CASUAL_OCTET_INVALID_HANDLE; 
+
+                  common::stream::write( std::cerr, "error: ", error, '\n');
 
```
Shall this be logged to log file or go to std::cerr? (Or possibly both?)

Noted that this construct (write to stderr) occurs in several places.

KristianIvarsson replied:
This was more a way to make the code/behaviour consistent with other C-buffers and the question is not just valid for this buffer type

KristianIvarsson committed:
buffer: fixed comments, error messages and changed behaviour of certain constructor

Some commented code was removed and a constructor used for null was changed to really mean null

lazan commented on middleware/buffer/source/octet.cpp:
```diff
@@ -173,13 +173,12 @@ namespace casual
             }
 
-         } // <unnamed>
+         } //
```
I think the `<unnamed>` gives information. Not just that someone has forgotten to comment the end of a namespace. But it's not that important.

lazan commented on middleware/buffer/source/order.cpp:
```diff
@@ -162,8 +163,12 @@ namespace casual
+                     if( error.code() == common::code::xatmi::argument)
+                        return CASUAL_ORDER_INVALID_HANDLE; 
+
+                     common::stream::write( std::cerr, "error: ", error, '\n');
```
Why do we write to `std::cerr` this could happen not only in CLI context I presume?

KristianIvarsson replied:
It's been like that for ages in CASUAL_FIELD, so it was just to make it coherent, but we can do whatever appropriate

lazan commented on middleware/buffer/source/string.cpp:
```diff
@@ -129,14 +129,22 @@ namespace casual
+                  if( error.code() == common::code::xatmi::argument)
+                     return CASUAL_STRING_INVALID_HANDLE; 
+
+                  common::stream::write( std::cerr, "error: ", error, '\n');
```
Same here...

KristianIvarsson replied:
Shall we log it into casual-log instead and if so, what category ?

lazan replied:
We've got the `common::log::error( <code>, <info>...)`. Logs to error category.  But what _code_ should you use? Check if you think you find something sufficient in `code::casual::???`. If not, maybe a `code::buffer` should be created?

lazan replied:
If you create one, I think it should be local to _buffer_. Should it not?

lazan replied:
Or, of course, you could add _codes_ in `code::casual` if you want.

lazan reviewed (commented):
Why log to `std::cerr`?

KristianIvarsson committed:
buffer: fixed error handling and comments

Writing certain error to log file instead of stderr and make sure that reallocating buffer with zero size doesn't disappear

KristianIvarsson committed:
buffer: simplified buffer allocation

The logic to make sure that data() for buffers returned a valid pointer existed in many many places and that was removed and now relies on the (central) Payload type

lazan commented:
Merged manually
